cpp/daal/src/algorithms/dtrees/forest/df_train_dense_default_impl.i

```diff
@@ -304,20 +304,59 @@ services::Status selectParallelizationTechnique(const Parameter & par, engines::
 //////////////////////////////////////////////////////////////////////////////////////////
 // compute() implementation
 //////////////////////////////////////////////////////////////////////////////////////////
-template <typename algorithmFPType, CpuType cpu, typename ModelType, typename TaskType>
+template <CpuType cpu, typename IndexType, typename BinIndexType>
+void computeImplDispatch(const size_t nRows, const size_t nCols, const IndexType * featureIndex,
+                         TVector<BinIndexType, cpu, ScalableAllocator<cpu> > & binIndexVector, BinIndexType ** binIndex);
+
+template <CpuType cpu, dtrees::internal::IndexedFeatures::IndexType, dtrees::internal::IndexedFeatures::IndexType>
+void computeImplDispatch(const size_t nRows, const size_t nCols, const dtrees::internal::IndexedFeatures::IndexType * featureIndex,
+                         TVector<dtrees::internal::IndexedFeatures::IndexType, cpu, ScalableAllocator<cpu> > & binIndexVector,
+                         dtrees::internal::IndexedFeatures::IndexType ** binIndex)
+{
+    *binIndex = const_cast<dtrees::internal::IndexedFeatures::IndexType *>(featureIndex);
+}
+
+template <CpuType cpu, typename IndexType, typename BinIndexType>
+void computeImplDispatch(const size_t nRows, const size_t nCols, const IndexType * featureIndex,
+                         TVector<BinIndexType, cpu, ScalableAllocator<cpu> > & binIndexVector, BinIndexType ** binIndex)
+{
+    binIndexVector.resize(nRows * nCols);
+    *binIndex = binIndexVector.get();
+
+    const size_t nThreads    = threader_get_threads_number();
+    const size_t nBlocks     = ((nThreads < nRows) ? nThreads : 1);
+    const size_t sizeOfBlock = nRows / nBlocks + !!(nRows % nBlocks);
+
+    daal::threader_for(nBlocks, nBlocks, [&](size_t iBlock) {
```

Comment:
use `conditional_threader_for` if for you important for you that under some conditions the code is compute sequentially.

Reply:
here only data copy, from my side we don't have to split array on many blocks